compiler/rustc_parse/src/parser/path.rs

```diff
@@ -624,9 +624,18 @@ impl<'a> Parser<'a> {
             GenericArg::Const(self.parse_const_arg()?)
         } else if self.check_type() {
             // Parse type argument.
+            let is_const_fn = self.look_ahead(1, |t| t.kind == token::OpenDelim(token::Paren));
+            let mut snapshot = self.diagnostic_snapshot();
             match self.parse_ty() {
                 Ok(ty) => GenericArg::Type(ty),
                 Err(err) => {
+                    if is_const_fn {
+                        if let Ok(expr) = snapshot.parse_expr_res(Restrictions::CONST_EXPR, None) {
+                            snapshot.unclosed_delims.extend(self.unclosed_delims.clone());
```

Comment:
This would do it but if anyone else notices `diagnotic_snapshot` and wants to use it, they might miss that this is required which could result in a bug later. You could maybe make a `SnapshotParser` struct containing the cloned `Parser` and the saved `unclosed_delims`, and then I think a `Deref` implementation to `Parser` let us use it like we would `self`, you could implement a function on `Parser` (or on `SnapshotParser`, the specifics aren't that important, as long as it is hard to misuse) that takes the `SnapshotParser` and replaces itself with it and restores the `unclosed_delims`. Ideally we'd be able to replace all of the other uses of `self.clone()` in the parser for diagnostics with this machinery, and we can put some documentation comments on it that describe how it is to be used, etc.